pkg/hubble/metrics/metric_config_watcher.go

```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package metrics
+
+import (
+	"context"
+	"crypto/md5"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"os"
+	"reflect"
+
+	"github.com/sirupsen/logrus"
+	"sigs.k8s.io/yaml"
+
+	"github.com/cilium/cilium/pkg/hubble/metrics/api"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+var metricReloadInterval = 10 * time.Second
+
+type metricConfigWatcher struct {
+	logger         logrus.FieldLogger
+	configFilePath string
+	callback       func(ctx context.Context, hash uint64, config api.Config)
+	ticker         *time.Ticker
+	stop           chan bool
+	currentCfgHash uint64
+	cfgStore       map[string]*api.MetricConfig
+	mutex          lock.RWMutex
+}
+
+// NewmetricConfigWatcher creates a config watcher instance. Config watcher notifies
+// DynamicFlowProcessor when config file changes and dynamic metric config should be
+// reconciled.
+func NewMetricConfigWatcher(
+	configFilePath string,
+	callback func(ctx context.Context, hash uint64, config api.Config),
+) *metricConfigWatcher {
+	watcher := &metricConfigWatcher{
+		logger:         logrus.New().WithField(logfields.LogSubsys, "hubble").WithField("configFilePath", configFilePath),
+		configFilePath: configFilePath,
+		callback:       callback,
+		currentCfgHash: 0,
+		cfgStore:       make(map[string]*api.MetricConfig),
+	}
+
+	// initial configuration load
+	watcher.reload()
+
+	watcher.ticker = time.NewTicker(metricReloadInterval)
+	watcher.stop = make(chan bool)
+
+	go func() {
+		for {
+			select {
+			case <-watcher.stop:
+				return
+			case <-watcher.ticker.C:
+				watcher.reload()
+			}
+		}
+	}()
+
+	return watcher
+}
+
+func (c *metricConfigWatcher) reload() {
+	c.logger.Debug("Attempting reload")
+	config, isSameHash, hash, err := c.readConfig()
+	if err != nil {
+		c.logger.WithError(err).Error("failed reading dynamic exporter config")
+	} else {
+		if !isSameHash {
+			c.callback(context.TODO(), hash, *config)
+		}
+	}
+}
+
+func (c *metricConfigWatcher) resetCfgPath(path string) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	c.configFilePath = path
+}
+
+// Stop stops watcher.
+func (c *metricConfigWatcher) Stop() {
+	if c.ticker != nil {
+		c.ticker.Stop()
+	}
+	c.stop <- true
```

Comment:
It probably makes sense to just close the channel here.

Reply:
If I do that, I'll need to check for the closure instead of consuming this value - is there much difference?
Taking a quick look here (https://gobyexample.com/closing-channels) this seems idiomatic (?)

Reply:
I am just realizing that the DynamicFlowProcessor is missing its `Stop` method that would call this one in response to a context closure event and stop the watcher.. Let me fix that up.
@chancez FYI Should we also `Deinit` all handlers there?

Reply:
We could add a Shutdown method to the `DynamicFlowExporter` to stop the watcher for sure. As far as running `Deinit` for all handlers in shutdown, I think that feels like a reasonable behavior.

Reply:
Cool, added in this commit: https://github.com/cilium/cilium/pull/35185/commits/2bda66ee05b35edbfa3920ac44d6416b20631568